src/ligero/mod.rs

```diff
@@ -0,0 +1,323 @@
+};
+use ark_ff::PrimeField;
+use ark_poly::Polynomial;
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use digest::Digest;
+
+use crate::{
+    Error, PCCommitment, PCCommitterKey, PCPreparedCommitment, PCPreparedVerifierKey, PCRandomness,
+    PCUniversalParams, PCVerifierKey, PolynomialCommitment,
+};
+
+use ark_std::rand::RngCore;
+
+// TODO: Disclaimer: no hiding prop
+/// The Ligero polynomial commitment scheme.
+pub struct Ligero<
+    F: PrimeField,
+    C: Config,
+    D: Digest,
+    S: CryptographicSponge,
+    /// one over the rate rho
+    const rho_inv: usize,
+    /// security parameter, used in calculating t
+    const sec_param: usize,
+> {
+    _field: PhantomData<F>,
+    _config: PhantomData<C>,
+    _digest: PhantomData<D>,
+    _sponge: PhantomData<S>,
+}
+
+// TODO come up with reasonable defaults
+const DEFAULT_RHO_INV: usize = 2;
+const DEFAULT_SEC_PARAM: usize = 128;
+
+fn calculate_t(rho_inv: usize, sec_param: usize) -> usize {
+    // TODO calculate t somehow
+    let t = 5;
+    t
+}
+
+impl<F, C, D, S, const rho_inv: usize, const sec_param: usize>
+    Ligero<F, C, D, S, rho_inv, sec_param>
+where
+    F: PrimeField + Borrow<<C as Config>::Leaf> + Absorb,
+    C: Config,
+    C::InnerDigest: Absorb,
+    D: Digest,
+    S: CryptographicSponge,
+{
+    /// Create a new instance of Ligero.
+    /// If either or both parameters are None, their default values are used.
+    pub fn new() -> Self {
+        Self {
+            _config: PhantomData,
+            _field: PhantomData,
+            _digest: PhantomData,
+            _sponge: PhantomData,
+        }
+    }
+
+    /// The verifier can check the well-formedness of the commitment by taking random linear combinations.
+    fn verify(
+        commitment: &LigeroPCCommitment<F, C>,
+        leaf_hash_params: &LeafParam<C>,
+        two_to_one_params: &TwoToOneParam<C>,
+        sponge: &mut S,
+    ) -> Result<(), Error> {
+        let t = calculate_t(rho_inv, sec_param);
+
+        // 1. Hash the received columns, to get the leaf hashes
+        let col_hashes: Vec<F> = vec![F::zero(); t];
+        for c in commitment.transcript.columns.iter() {
+            // TODO some hashing, with the digest?
+        }
+
+        // 2. Verify the paths for each of the leaf hashes
+        // TODO need a way to relate the index to the leaf hash
+        for (i, leaf) in col_hashes.iter().enumerate() {
+            // TODO handle the error here
+            commitment.transcript.paths[i]
+                .verify(
+                    leaf_hash_params,
+                    two_to_one_params,
+                    &commitment.root,
+                    leaf.clone(),
+                )
+                .unwrap();
+        }
+
+        // 3. verify the random linear combinations
+
+        // 4. Verify the Fiat-Shamir transformation
+        sponge.absorb(&commitment.root);
+        let verifiers_randomness: Vec<F> = sponge.squeeze_field_elements(t);
```

Comment:
I wonder if one single element (the root) generates enough randomness to squeeze t elements, which could be many. But I can't see other good sources of initial randomness atm.

Reply:
I think that's the only option indeed